File: 056/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <math.h>
using namespace std;

struct Interval
{
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};

class Solution
{
  public:
    static bool cmp(Interval a, Interval b)
    {
        if (a.start != b.start)
        {
            return a.start < b.start;
        }
        else
        {
            return a.end < b.end;
        }
    }
    vector<Interval> merge(vector<Interval> &intervals)
    {
        vector<Interval> v;
        if (intervals.empty())
            return v;
        sort(intervals.begin(), intervals.end(), cmp);
        for (int i = 0; i < intervals.size() - 1; i++)
        {
            if (intervals[i].end < intervals[i + 1].start)
            {
                v.push_back(intervals[i]);
            }
            else
            {
                intervals[i + 1].start = intervals[i].start;
                intervals[i + 1].end = max(intervals[i].end, intervals[i + 1].end);
            }
        }
        v.push_back(intervals[intervals.size() - 1]);
        return v;
    }
};

int main()
{
    Solution sol;
    vector<Interval> intervals;
    struct Interval s1(1, 3);
    struct Interval s2(2, 6);
    struct Interval s3(8, 10);
    struct Interval s4(15, 18);
    intervals.push_back(s1);
    intervals.push_back(s2);
    intervals.push_back(s3);
    intervals.push_back(s4);
    vector<Interval> v;
    v = sol.merge(intervals);
    for (int i = 0; i < v.size(); i++)
    {
        cout << v[i].start << " " << v[i].end << endl;
    }
    return 0;
}

File: 056/main.js
/**
 * Definition for an interval.
 * function Interval(start, end) {
 *     this.start = start;
 *     this.end = end;
 * }
 */
/**
 * @param {Interval[]} intervals
 * @return {Interval[]}
 */

function Interval (start, end) {
  this.start = start
  this.end = end
}

var merge = function (intervals) {
  const res = []
  if (intervals == null || intervals.length === 0) return res
  intervals.sort((a, b) => {
    if (a.start !== b.start) {
      return a.start - b.start
    } else {
      return a.end - b.end
    }
  })
  for (let i = 0; i < intervals.length - 1; i++) {
    if (intervals[i].end < intervals[i + 1].start) {
      res.push(intervals[i])
    } else {
      intervals[i + 1].start = intervals[i].start
      intervals[i + 1].end = Math.max(intervals[i].end, intervals[i + 1].end)
    }
  }
  res.push(intervals[intervals.length - 1])
  return res
}

const l1 = new Interval(1, 3)
const l2 = new Interval(2, 6)
const l3 = new Interval(8, 10)
const l4 = new Interval(15, 18)
let i1 = [l1, l2, l3, l4]
console.log(merge(i1))
